src/storage.rs

```diff
@@ -686,6 +692,155 @@ impl ArchetypeData {
         self.tags.validate(self.chunk_sets.len());
     }
 
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn clone_merge<C: crate::world::CloneMergeImpl>(
+        &mut self,
+        src_world: &crate::world::World,
+        src_archetype: &ArchetypeData,
+        dst_archetype_index: usize,
+        dst_entity_allocator: &mut crate::entity::EntityAllocator,
+        dst_resources: &crate::resource::Resources,
+        clone_impl: &C,
+        replace_mappings: Option<&std::collections::HashMap<Entity, Entity>>,
```

Comment:
Would it be possible to generalize or genericize the hashmaps and collections coming into these functions? Forcing the std hashmaps and their hashers kind of sucks; at a minimum, it'd be nice to have it generic over the hasher. Ideally, it'd be more flexible to somehow do this iter based if we could; but that would lose a lot of lookup value.

Reply:
One option here would be to defer to the passed-in trait impl. So we wouldn't pass anything in at all and instead hit functions on the trait impl to query for lookups and notify of the resulting mappings.

I'm not sure what the performance implications would be (hopefully it gets inlined since this isn't a virtual call?) and there's slightly more risk of an implementor doing something bad.. for example not returning consistent results during the full lifetime of the call.